[d3-axis and d3-format] activate strictNullChecks and strictFunctionTypes

denisname opened the pull request:
In this PR, I activated strictFunctionTypes in d3-axis. To get it work I also needed to update d3-format.

In d3-format:
- Activate strictNullChecks in d3-format, just needed to change FormatSpecifier#precision.
- Use type `{ valueOf(): number; }` in `format` and `formatPrefix`.
- Function `format` can take a string, see [format-type-c-test.js](https://github.com/d3/d3-format/blob/master/test/format-type-c-test.js).
- Add `@throws` in JsDoc.

In d3-axis:
- Remove `"unified-signatures": false` from `tslint.json`.
- Use `$ExpectError` for failing tests.
- Activate strictFunctionTypes in d3-axis: change `Selection<SVGSVGElement | SVGGElement, any, any, any>` by `Selection<SVGSVGElement, any, any, any> | Selection<SVGGElement, any, any, any>`.

I add a test for `strictFunctionTypes`. Not sure if it is the best way to do it. Also CI build fail for them because some `$ExpectError` lines are not detected as failing. I'm wondering if `dtslint` test the code with [TS 2.7](https://github.com/Microsoft/definitelytyped-header-parser/blob/master/index.ts#L13) or not.

Finally, I don't think any new features from TS 2.3 may improve those definitions. So I kept the minimum version to 2.0.

Related: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/23611.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [d3-format](https://github.com/d3/d3-format/blob/master/README.md) and [d3-axis](https://github.com/d3/d3-axis/blob/master/README.md)
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

tomwanzek commented:
@denisname Thank you for submitting this PR. Given the extent of the contemplated changes, I will take some time to review them. In particular, since at first glance, you are proposing diverting from some of the conventions we have used in writing previous definitions. I will explain in my comments.

I will take the time to digest and respond with motivations for some of the conventions we followed in the past. Should we reconsider some of them, I would prefer to have full alignment first and then update other definitions as appropriate for consistency.

So that being said, it might be worthwhile to hold off on your proposed efforts with **d3-array**, until we have such alignment. This will save cycles for you and reviewers, should there be items where we prefer to hold on to current conventions. With 30+ definitions in the core D3 universe, conventions matter for a consistent user experience.

Thanks in advance for your understanding.

denisname commented:
Ok. Note that I cut my changes into several small commits to make revision simpler.

tomwanzek commented on types/d3-format/index.d.ts:
```diff
@@ -44,20 +44,22 @@ export interface FormatLocaleObject {
      */
-    format(specifier: string): (n: number) => string;
+    format(specifier: string): (n: { valueOf(): number; } | string) => string;
 
```
The matter of specifier type "c" is interesting in that the official API doc does not reflect the implemented code since [this change to the d3-format code](https://github.com/d3/d3-format/commit/e247671b89d78fa27106bafede4d7703c5ac1eb0). The official API doc still refers to the old behaviour:
```
c - converts the integer to the corresponding unicode character before printing.
```
and the reference to the argument of the returned formatter function being a "number". So good on you for finding a rather undocumented behaviour in the source.

Seems to me on yet further consideration, that (a) the API doc should be updated to reflect the current behaviour (PR to d3-format) AND (b) we need a more complex explanation of the permissible inputs to the formatter functions as well, i.e. when an actual text string is permissible in combination with format-specifiers of type "c".

So, I stand corrected on the extension to also include `string` explicitly. No-one seems to have used it so far in actual applications with these TS definitions, otherwise we would have likely heard about it.

tomwanzek commented on types/d3-axis/tslint.json:
```diff
@@ -1,7 +1,6 @@
     "extends": "dtslint/dt.json",
     "rules": {
-        "unified-signatures": false,
-        "no-unnecessary-generics": false
```
This will need to be added back in, in light of the `scale<A ...>(): A` signature.

tomwanzek commented on types/d3-axis/d3-axis-tests.ts:
```diff
@@ -76,15 +78,18 @@ let leftAxis: d3Axis.Axis<number | { valueOf(): number }> = d3Axis.axisLeft(scal
 const axisScale: d3Axis.AxisScale<string> = bottomAxis.scale();
-const ordinalScale: ScaleOrdinal<string, number> = bottomAxis.scale<ScaleOrdinal<string, number>>();
-// ordinalScale = bottomAxis.scale(); // fails, without casting as AxisScale is purposely  generic
+const ordinalScale: ScaleOrdinal<string, number> = bottomAxis.scale() as ScaleOrdinal<string, number>;
```
Use declarative casting`bottomAxis.scale<ScaleOrdinal<string, number>>()`

tomwanzek commented on types/d3-axis/d3-axis-tests.ts:
```diff
@@ -76,15 +78,18 @@ let leftAxis: d3Axis.Axis<number | { valueOf(): number }> = d3Axis.axisLeft(scal
 leftAxis = leftAxis.scale(scalePow());
-const powerScale: ScalePower<number, number> = leftAxis.scale<ScalePower<number, number>>();
-// powerScale = leftAxis.scale(); // fails, without casting as AxisScale is purposely  generic
+const powerScale: ScalePower<number, number> = leftAxis.scale() as ScalePower<number, number>;
```
Use declarative casting`leftAxis.scale<ScalePower<number, number>>()`

tomwanzek commented on types/d3-axis/d3-axis-tests.ts:
```diff
@@ -149,14 +155,52 @@ const gTransition = gSelection.transition();
+
+// --------------------------------------------------------------------------
+// Strict Function Types
+// --------------------------------------------------------------------------
```
Laudable idea to think of explicitly testing `strictFunctionTypes`. However, it will not work, as you found, since dts-lint is running from the minimally permitted TS version up to the latest version. So for anything below TS 2.6 the tests will not fail as expected, as `strictFunctionTypes` were not enforced.

So I would simply remove this section.

tomwanzek commented on types/d3-axis/d3-axis-tests.ts:
```diff
@@ -58,7 +59,8 @@ const canvas: HTMLCanvasElement = select<HTMLCanvasElement, any>('canvas').node(
 containerElement = svg;
 containerElement = g;
-// containerElement = canvas; // fails, incompatible type
+// $ExpectError
```
Thanks for starting to use this dts-lint feature. We should start using this across the board on all the "fail" tests in D3. We had to historically comment them out (and manually check during definition updates). So let's roll with `$ExpectError`

tomwanzek commented on types/d3-axis/d3-axis-tests.ts:
```diff
@@ -149,14 +155,52 @@ const gTransition = gSelection.transition();
 gTransition.call(topAxis);
 
-const svgSelection: Selection<SVGSVGElement, any, any, any> = select<SVGSVGElement, any>('g');
+const svgSelection: Selection<SVGSVGElement, any, any, any> = select<SVGSVGElement, any>('svg');
```
Good catch 😄 Now that's attention to detail!

tomwanzek commented on types/d3-axis/index.d.ts:
```diff
@@ -29,53 +29,53 @@ export interface AxisTimeInterval {
- * Interface defining an axis generator. The generic <Domain> is the type of the axis domain
+ * Interface defining an axis generator. The generic <Domain> is the type of the axis domain.
  */
 export interface Axis<Domain> {
```
I will review the constraints on `Domain` in the definition for `Axis` and `AxisScale` to define a suitable default for the generic parameters (TS 2.3 feature that we should use here as well). We might be able to do better than `any` for the most general permissible case.

We could have `export interface Axis<Domain extends DomainConstraint = DomainDefault> {...}` and `export interface AxisScale<Domain extends DomainConstraint = DomainDefault> {...}`. Then in day-to-day use one could just do `let x: Axis;` for brevity, if the defaults are good enough.

tomwanzek commented:
As for a minimal constraint on the generic `Domain`, we could define:
```ts
export type AxisDomain = number | string | Date | { valueOf(): number};
```
and then impose:
```ts
export interface AxisScale<Domain extends AxisDomain> {...}
...
export interface Axis<Domain extends AxisDomain> { ... }
...
export function axisTop<Domain extends AxisDomain>(scale: AxisScale<Domain>): Axis<Domain>;
export function axisRight<Domain extends AxisDomain>(scale: AxisScale<Domain>): Axis<Domain>;
export function axisBottom<Domain extends AxisDomain>(scale: AxisScale<Domain>): Axis<Domain>;
export function axisLeft<Domain extends AxisDomain>(scale: AxisScale<Domain>): Axis<Domain>;
```

Defaulting the generic for `Axis` and `AxisScale` to `AxisDomain` turns out to be impractical due to the impact of `strictFunctionChecks`.

tomwanzek commented:
@denisname Penny for your thoughts...:smile:

denisname commented:
- numerals?: [string, string, ..., string]
  Fair point about assigning a string[]-inferred-type.
- format[Prefix]
  Let's go for `(n: number | { valueOf(): number }) => string`.
- strict function types test
  Removed!
- scale generic:
  While making the changes, I realize that the tests line [80](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2e3cfeee51f0fcf6efda97c9453aa4946ae77fca/types/d3-axis/d3-axis-tests.ts#L80) and [87](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2e3cfeee51f0fcf6efda97c9453aa4946ae77fca/types/d3-axis/d3-axis-tests.ts#L87) are no failing as expected. The errors are caused by reassigning a `const` and not because there is a generic missing. The expected error stop happening with [TS 2.4](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-4.html) which "improved inference for generics". I'm not sure what to do with the two tests. Best is probably to remove them!?

  I take this opportunity to say also that I disagree about `scale` being generic. See [getMeAT](https://github.com/DefinitelyTyped/DefinitelyTyped/#common-mistakes) for reasons. Also, if you don't like `as` keyword, `<ScaleOrdinal<string, number>>bottomAxis.scale()` is same number of characters than `bottomAxis.scale<ScaleOrdinal<string, number>>()`. One more with a space. But, I recognise the importance of keeping the different d3 definitions consistent and to avoid breaking backward compatibility. And as TS 2.4 makes the generic "optional", I'm going to set the generic of `scale` back…

denisname commented:
The `AxisDomain` constraint seems good. 👍

I also tried to solve the problem of [ticks et ticksFormat typing](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/d3-axis/index.d.ts#L40-L44) and to remove: [`tickArguments(args: any[]): this;`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/d3-axis/index.d.ts#L183). Because of that overloading, any array is accepted by the compiler for `tickArguments`...

For this, I separated `AxisScale` in two (see [full code](https://gist.github.com/denisname/40f2d711b5280da4f2d56a17bf47b8c4))

```typescript
export interface AxisScale<Domain> {
    (x: Domain): number | undefined;
    domain(): Domain[];
    range(): number[];
    copy(): this;
    bandwidth?(): number;
}

export interface TickedAxisScale<Domain, Tick> extends AxisScale<Domain> {
    ticks(count?: Tick): Domain[];
    tickFormat(count?: Tick, specifier?: string): ((d: Domain) => string);
}
```

I then removed `ticks(.)` and `ticksFormat(.)` from `Axis<Domain>`, and added:

```typescript
export interface TickedAxis<Domain, Tick> extends Axis<Domain> {
    // ...
    ticks(count: Tick, specifier?: string): this;
    tickArguments(): [Tick] | [Tick, string];
    tickArguments(args: [Tick] | [Tick, string]): this;
}

export function axisTop<Domain, Tick>(scale: TickedAxisScale<Domain, Tick>): TickedAxis<Domain, Tick>;
export function axisRight<Domain, Tick>(scale: TickedAxisScale<Domain, Tick>): TickedAxis<Domain, Tick>;
export function axisBottom<Domain, Tick>(scale: TickedAxisScale<Domain, Tick>): TickedAxis<Domain, Tick>;
export function axisLeft<Domain, Tick>(scale: TickedAxisScale<Domain, Tick>): TickedAxis<Domain, Tick>;
```

Now type inference detect which types are allowed for `ticks(.)` and `ticksFormat(.)`, see [gist](https://gist.github.com/denisname/3bbe77986862a5009a956ab4f8b3dc3e) for all tests. For example:

```typescript
d3Axis.axisTop(scaleLinear()).ticks(1, ',f'); // Ok
d3Axis.axisTop(scaleLinear()).ticks(timeMinute.every(5)!, ',f'); // fails
```

Provided that, in `d3-scale`, the two [`ticks(.)`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/42d220f6611ba11975c9aab84f9b1c0c027fcda6/types/d3-scale/index.d.ts#L841-L852) are merged: `ticks(count: number | TimeInterval): Date[];`.

In the gist I also added `TickedCustomAxisScale`. But, are there really any people who create scales that have a singnature for `ticks` and` ticksFormat` other than `number` or` TimeInterval`!? And if that's the case, is it not better for them to define their own adapted interface, rather than using `TickedCustomAxisScale`?

Hope my explanations are clear. Waiting for your comments. :)

tomwanzek commented:
@denisname Thanks, I will respond at my earliest. Unfortunately, I am little otherwise-committed right now :smile:

mhegazy commented:
Thanks for your contribution. This PR has failing CI tests and can not be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.

tomwanzek commented:
PS: Ignore the comment suggesting to add the default to the generic in `scale<...>()`. Of course, that is not needed, as the constraint will form the default in this case.

As for the lines that create problems for `// $ExpectError`, for one, if a `const` should  be a `let`, let's fix that. That was the issue with `let` creating a linter error, if the re-assignment line is commented out.

If we are still running into issues where the error assertion does not work in some TS version, but works in others, maybe we should just remove that particular definitions test.
